internal/databases/postgres/snapshot_push_handler.go

```diff
@@ -0,0 +1,332 @@
+	backupLabel   string
+	tablespaceMap string
+}
+
+// NewSnapshotBackupArguments creates a SnapshotBackupArguments object
+func NewSnapshotBackupArguments(uploader internal.Uploader, pgDataDirectory string, backupsFolder string,
+	isPermanent bool, userData interface{}, snapshotCommand string) SnapshotBackupArguments {
+	return SnapshotBackupArguments{
+		Uploader:        uploader,
+		pgDataDirectory: pgDataDirectory,
+		backupsFolder:   backupsFolder,
+		isPermanent:     isPermanent,
+		userData:        userData,
+		snapshotCommand: snapshotCommand,
+	}
+}
+
+// NewSnapshotBackupHandler creates a new SnapshotBackupHandler
+func NewSnapshotBackupHandler(arguments SnapshotBackupArguments) (sbh *SnapshotBackupHandler, err error) {
+	// Get PostgreSQL server info
+	pgInfo, _, err := GetPgServerInfo(false)
+	if err != nil {
+		return nil, err
+	}
+
+	sbh = &SnapshotBackupHandler{
+		Arguments: arguments,
+		PgInfo:    pgInfo,
+	}
+
+	return sbh, nil
+}
+
+// HandleSnapshotPush handles the snapshot backup process
+func (sbh *SnapshotBackupHandler) HandleSnapshotPush(ctx context.Context) {
+	sbh.CurBackupInfo.StartTime = utility.TimeNowCrossPlatformUTC()
+
+	if sbh.Arguments.pgDataDirectory == "" {
+		tracelog.ErrorLogger.Fatal("PGDATA must be specified for snapshot backups")
+	}
+
+	// Verify data directory matches
+	fromCli := sbh.Arguments.pgDataDirectory
+	fromServer := sbh.PgInfo.PgDataDirectory
+	if utility.AbsResolveSymlink(fromCli) != fromServer {
+		tracelog.ErrorLogger.Fatalf("Data directory from command line '%s' is not the same as Postgres' one '%s'",
+			fromCli, fromServer)
+	}
+
+	sbh.createAndPushSnapshotBackup(ctx)
+}
+
+func (sbh *SnapshotBackupHandler) createAndPushSnapshotBackup(ctx context.Context) {
+	// Change to backups folder
+	sbh.Arguments.Uploader.ChangeDirectory(sbh.Arguments.backupsFolder)
+	tracelog.DebugLogger.Printf("Uploading folder: %s", sbh.Arguments.Uploader.Folder())
+
+	// Step 1: Connect to postgres and start backup
+	err := sbh.startBackup()
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	// Step 2: Execute the snapshot command
+	err = sbh.executeSnapshotCommand()
+	if err != nil {
+		// If snapshot command fails, we must stop the backup gracefully
+		sbh.stopBackupOnError()
+		tracelog.ErrorLogger.FatalOnError(err)
+	}
+
+	// Step 3: Stop backup and get finish LSN
+	err = sbh.stopBackup()
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	// Step 4: Upload sentinel metadata
+	sbh.uploadSnapshotMetadata(ctx)
+
+	tracelog.InfoLogger.Printf("Wrote snapshot backup with name %s", sbh.CurBackupInfo.Name)
+}
+
+func (sbh *SnapshotBackupHandler) startBackup() error {
+	// Connect to postgres
+	tracelog.DebugLogger.Println("Connecting to Postgres for snapshot backup.")
+	conn, err := Connect()
+	if err != nil {
+		return err
+	}
+
+	sbh.QueryRunner, err = NewPgQueryRunner(conn)
+	if err != nil {
+		return errors.Wrap(err, "failed to build query runner")
```

Comment:
NIT: it seems that `github/pkg/errors` is public archive and we can use `fmt.Errorf()` instead.